### PR TITLE
GROUP-BY prioritizes input columns in case of ambiguity

### LINKED ISSUE #9162: Probably incorrect result when the column referenced in GROUP BY is ambiguous
### Describe the bug

The result might be incorrect when the column referenced in the GROUP BY clause is ambiguous; that is, when the column exists in both the select list and the input. This behavior is different from PostgreSQL.

### To Reproduce

Run the following queries in DataFusion CLI .
```sh
DataFusion CLI v35.0.0
❯ create table t(a bigint);
0 rows in set. Query took 0.009 seconds.

❯ insert into t values(1),(2),(3);
+-------+
| count |
+-------+
| 3     |
+-------+
1 row in set. Query took 0.007 seconds.

❯ select 0 as "t.a" from t group by t.a;
+-----+
| t.a |
+-----+
| 0   |
+-----+
1 row in set. Query took 0.010 seconds.
```

### Expected behavior

The result should ideally be the same as PostgreSQL, MySQL, and DuckDB.
```
psql => select 0 as "t.a" from t group by t.a;
 t.a
-----
   0
   0
   0
(3 rows)
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jonahgao**
## Which issue does this PR close?
Closes #9162.

## Rationale for this change

When a column referenced by group-by exists both in the select list and the input, the one from the input should be given priority. In issue 9162, there are two references with the same name: one is an unqualified "t.a," and the other is a qualified t.a.

This is the practice of many databases, including PostgreSQL, Oracle, MySQL, Duckdb, etc. 
In the PostgreSQL documentation, there is an [explanation](https://www.postgresql.org/docs/current/sql-select.html#SQL-GROUPBY) about it.
> An expression used inside a grouping_element can be an input column name, or the name or ordinal number of an output column (SELECT list item), or an arbitrary expression formed from input-column values. In case of ambiguity, a GROUP BY name will be interpreted as an **input-column** name rather than an output column name.

## What changes are included in this PR?
Prioritize searching the schema of the base plan when generating GROUP BY expressions.

## Are these changes tested?
Yes

## Are there any user-facing changes?
No

### CODE REVIEW

File: datafusion/sqllogictest/test_files/aggregate.slt [hunk 1]
```diff
@@ -3179,3 +3179,23 @@ NULL
 statement ok
 DROP TABLE t;
 
+
+# Test for the case when the column name is ambiguous
+statement ok
+CREATE TABLE t(a BIGINT) AS VALUES(1), (2), (3);
+
+# The column name referenced by GROUP-BY is ambiguous
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/aggregate.slt:3187]:
```suggestion
# The column name referenced by GROUP-BY is ambiguous, prefer the column in base plan
```

File: datafusion/sqllogictest/test_files/aggregate.slt [hunk 2]
```diff
@@ -3179,3 +3179,23 @@ NULL
...
+query I
+SELECT 0 as "t.a" FROM t GROUP BY t.a;
+----
+0
+0
+0
+
+# The column name referenced by HAVING is ambiguous
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/aggregate.slt:3195]:
```suggestion
# The column name referenced by HAVING is ambiguous, prefer the column in the base plan
```

**Review by alamb — APPROVED** [review]
Thank you @jonahgao  -- this change makes sense to me

I believe we have seen issues related to this in InfluxDB as well. 

I verified that this PR makes DataFusion consistent with postgres

```sql
postgres=# CREATE TABLE t(a BIGINT);
CREATE TABLE
postgres=# INSERT INTO t  VALUES (1), (2), (3);
INSERT 0 3
postgres=# SELECT 0 as "t.a" FROM t GROUP BY t.a;
 t.a
-----
   0
   0
   0
(3 rows)

postgres=# SELECT 0 AS "t.a" FROM t HAVING MAX(t.a) = 0;
 t.a
-----
(0 rows)

postgres=#
```

I also verified on main that DataFusion takes the values from the select list

```shell
andrewlamb@Andrews-MacBook-Pro:~/Software/arrow-datafusion$ datafusion-cli
DataFusion CLI v35.0.0
❯ CREATE TABLE t(a BIGINT) AS VALUES(1), (2), (3);
0 rows in set. Query took 0.011 seconds.

❯ SELECT 0 as "t.a" FROM t GROUP BY t.a;
+-----+
| t.a |
+-----+
| 0   |
+-----+
1 row in set. Query took 0.007 seconds.

❯ SELECT 0 AS "t.a" FROM t HAVING MAX(t.a) = 0;
+-----+
| t.a |
+-----+
| 0   |
+-----+
1 row in set. Query took 0.004 seconds.
```

**Comment by jonahgao** [comment]
Thank you @alamb for reviewing and for the suggestions.

**Comment by alamb** [comment]
Thanks again @jonahgao

**Pull request merged**